File: Data Structures & Algorithms/interleaving-string/submission-3.kt
class Solution {
    fun isInterleave(s1: String, s2: String, s3: String): Boolean {
        if (s1.length + s2.length != s3.length)
            return false

        val m = s1.length + 1
        val n = s2.length + 1
        val grid = Array(m) { BooleanArray(n) {false} }
        grid[0][0] = true

        for (i in 1 until m)
            grid[i][0] = grid[i-1][0] && s1[i-1] == s3[i-1]

        for (i in 1 until n)
            grid[0][i] = grid[0][i-1] && s2[i-1] == s3[i-1]

        for (i in 1 until m)
            for (j in 1 until n)
                grid[i][j] = (grid[i-1][j] && s1[i-1] == s3[i+j-1]) || (grid[i][j-1] && s2[j-1] == s3[i+j-1])
        
        return grid[m-1][n-1]
    }
}

File: Data Structures & Algorithms/swim-in-rising-water/submission-3.py
class Solution:
    def swimInWater(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        heap = []
        dir = [(1, 0), (0, 1), (0, -1), (-1, 0)]

        heapq.heappush(heap, (grid[0][0], 0, 0))
        visited = [[False]*n for _ in range(m)]
        visited[0][0] = True
        while heap:
            height, i, j = heapq.heappop(heap)

            if i == m -1 and j == n-1:
                return height
            
            for d in dir:
                ni, nj = i + d[0], j + d[1]

                if 0 <= ni < m and 0 <= nj < n and not visited[ni][nj]:
                    visited[ni][nj] = True
                    heapq.heappush(heap, (max(height, grid[ni][nj]), ni, nj))
        
        return 0